### PR TITLE
Add nested `AGENTS.md` files for AI agent codebase guidance

### PR DESCRIPTION
**Pull request by strickvl**
## Summary

This PR introduces a distributed set of `AGENTS.md` files across the ZenML codebase to capture implicit development knowledge for AI coding agents (Claude Code, Cursor, Codex, etc.). The content is derived from a knowledge transfer session with @schustmi to document non-obvious patterns, gotchas, and best practices that typically surface during PR reviews.

## Motivation

AI coding agents are increasingly being used to contribute to the ZenML codebase. These agents benefit from explicit documentation of patterns that human developers learn through experience and code review feedback. By placing guidance files in relevant directories, agents discover context-appropriate rules when working in specific areas.

### New Files Created (3)

| File | Purpose |
|------|---------|
| `src/zenml/integrations/AGENTS.md` | **Critical:** Flavor file import rules (the #1 review concern for integration PRs), dependency update guidance, breaking change definitions |
| `src/zenml/orchestrators/AGENTS.md` | Deep-dive on `get_orchestrator_run_id` implementation (static vs dynamic pipelines), Kubernetes exception, reference implementations |
| `src/zenml/zen_server/AGENTS.md` | Server import isolation rule — code outside `zen_server/` should never import from within it |

### Files Updated (5)

| File | Additions |
|------|-----------|
| `AGENTS.md` (root) | Utils organization guidance, private methods & API stability, PR reviewer checklist gateway, migration testing workflow |
| `CLAUDE.md` | Ported crucial sections from AGENTS.md (utils placement, private methods, PR checklist) |
| `src/zenml/models/AGENTS.md` | Model changes backwards compatibility table, client method patterns, CLI-client filter coupling warning |
| `src/zenml/cli/AGENTS.md` | CLI import considerations (avoid integration library imports at module level) |
| `src/zenml/zen_stores/schemas/AGENTS.md` | Expanded scope to cover `zen_stores/` directory, SQL import restrictions, migration testing workflow, database column limits (~250 chars) |

## Key Patterns Documented

1. **Integration Flavor Import Rule** — Never import integration libraries at top level in flavor files (breaks ZenML if integration not installed)
2. **`get_orchestrator_run_id` Contract** — Must be unique per run AND same for all steps (tricky for static pipelines)
3. **CLI-Client Filter Coupling** — Adding filter model fields requires updating client method signatures
4. **Import Boundaries** — `zen_server/` and `zen_stores/` should not be imported from outside their directories
5. **Model Backwards Compatibility** — Adding properties is safe; deleting/renaming is breaking for server/client compat
6. **Private Methods** — Underscore-prefixed methods have no backwards compat guarantees; integrations should never use them

## Progressive Disclosure Structure

```
AGENTS.md (root)
├── General code style, testing, workflow, PR checklist gateway
│
├── src/zenml/integrations/AGENTS.md
│   └── Integration development: flavor imports, dependency updates
│
├── src/zenml/orchestrators/AGENTS.md
│   └── Orchestrator implementation: get_orchestrator_run_id, reference examples
│
├── src/zenml/zen_server/AGENTS.md
│   └── Server import isolation: don't import zen_server from outside
│
├── src/zenml/models/AGENTS.md
│   └── Domain models, filter-CLI coupling, backwards compatibility
│
├── src/zenml/cli/AGENTS.md
│   └── CLI patterns, list commands, import restrictions
│
└── src/zenml/zen_stores/schemas/AGENTS.md
    └── ORM schemas, migrations, SQL access patterns, column limits
```

## Test Plan

- [x] Verified all AGENTS.md files exist and contain expected content
- [x] Cross-referenced against original knowledge session document to ensure completeness
- [x] Files follow existing documentation patterns (markdown tables, code examples, checklists)

## Notes

- These files are intended for AI agents but are also useful for human contributors
- The root `AGENTS.md` and `CLAUDE.md` include a PR reviewer checklist that links to detailed explanations in nested files
- Content derived from live knowledge transfer session (2025-11-27)

### CODE REVIEW

**Comment by strickvl** [comment]
@htahir1 FYI

File: src/zenml/orchestrators/AGENTS.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+## Implementation Checklist
+
+When implementing a new orchestrator:
+
+- [ ] Inherit from `ContainerizedOrchestrator` if your orchestrator runs steps in containers
+- [ ] Implement `get_orchestrator_run_id()` following the static/dynamic patterns above
+- [ ] Implement `submit_pipeline()` for static pipelines
+- [ ] Optionally implement `submit_dynamic_pipeline()` for dynamic pipeline support
+- [ ] Handle `deployment.schedule` if your backend supports scheduling
```
Comment (schustmi) [inline src/zenml/orchestrators/AGENTS.md]: `snapshot.schedule`. Also, one could add that the resource settings should also be handled appropriately

File: src/zenml/zen_stores/schemas/AGENTS.md [hunk 2]
```diff
@@ -1,4 +1,80 @@
...
+4. **Test whether the migration works** — run `alembic upgrade head`
+
+**Why this matters:** Migrations must handle existing data from older versions. Testing only with fresh databases misses edge cases.
+
+**Note:** CI performs basic migration testing by creating runs on all versions, but local testing with realistic data is still recommended for complex migrations.
+
+### Migration Best Practices
+
+- Create migrations with descriptive names: `alembic revision -m "Add X to Y table"`
+- Test both upgrade and downgrade: `alembic upgrade head` and `alembic downgrade -1`
```
Comment (schustmi) [inline src/zenml/zen_stores/schemas/AGENTS.md]: Downgrade we don't actually support in most cases, so this is optional I think.

File: AGENTS.md [hunk 3]
```diff
@@ -68,6 +68,87 @@ Use filesystem navigation tools to explore the codebase structure as needed.
...
+**When changing private methods:**
+1. Search for usages **within the ZenML codebase** (grep/find references)
+2. Update all internal usages
+3. No need to worry about external user backwards compatibility
+
+**Critical rule for integrations:**
+
+> ⚠️ **Integrations should NEVER use ZenML private methods**
+
+When integrations move out of the main ZenML repo (external packages), mypy won't detect if a private method they depend on was changed. This leads to silent breakage. Always use public APIs in integration code.
```
Comment (schustmi) [inline AGENTS.md]: I guess this is more of a future concern? Anyway, it's a good practice to not use any private methods

File: AGENTS.md [hunk 4]
```diff
@@ -68,6 +68,87 @@ Use filesystem navigation tools to explore the codebase structure as needed.
...
+**The rule:**
+- `_method()` on a class → only call from within that class
+- `_function()` in a utils module → only call from within that module
+- This isn't always consistently applied in the codebase, but it's the intended convention
+
+**Backwards compatibility implications:**
+
+| Symbol type | Part of public API? | Breaking change if modified? |
+|-------------|---------------------|------------------------------|
+| `public_method()` | ✅ Yes | ⚠️ Yes — requires deprecation |
```
Comment (schustmi) [inline AGENTS.md]: I think this is a little too harsh. Changing the signature or behaviour of any function/method in our code base that doesn't start with an underscore is not necessarily a breaking change.

As we don't really have any written rules on this, it needs to be judged on a case-by-case basis.
- If it's some internal method that is used very deep inside our code, then this isn't intended for users and therefore can be changed, as long as all usages within our codebase are accounted for.
- If a method/class is imported in `zenml.__init__` (the root module), that for sure is part of the public API and therefore it will be a breaking change. Also applies to any public methods on those classes.
